core/lib/zksync_core/src/eth_sender/l1_batch_commit_data_generator.rs

```diff
@@ -0,0 +1,89 @@
+use zksync_l1_contract_interface::{
+    i_executor::{
+        methods::{CommitBatchesRollup, CommitBatchesValidium},
+        structures::{CommitBatchInfoRollup, CommitBatchInfoValidium},
+    },
+    Tokenizable, Tokenize,
+};
+use zksync_types::{commitment::L1BatchWithMetadata, ethabi::Token, pubdata_da::PubdataDA};
+
+/// [`L1BatchCommitDataGenerator`] abstracts how a batch or list of batches need to be tokenized
+/// for committing to L1.
+pub trait L1BatchCommitDataGenerator
+where
+    Self: std::fmt::Debug + Send + Sync,
+{
+    /// [`l1_commit_batches`] is the method you want when committing a block.
+    /// It uses metadata from `last_committed_l1_batch` and the batches in `l1_batches` to produce
+    /// the full commit data.
+    fn l1_commit_batches(
+        &self,
+        last_committed_l1_batch: &L1BatchWithMetadata,
+        l1_batches: &[L1BatchWithMetadata],
+        pubdata_da: &PubdataDA,
+    ) -> Vec<Token>;
+
+    /// [`l1_commit_batch`] is used mostly for size calculations for sealing criteria and for
+    /// consistency checks. Instead of preparing a full commit, it will tokenize an individual batch.
+    fn l1_commit_batch(&self, l1_batch: &L1BatchWithMetadata, pubdata_da: &PubdataDA) -> Token;
+}
+
+/// [`RollupModeL1BatchCommitDataGenerator`] implements [`L1BatchCommitDataGenerator`] for
+/// contracts operating in rollup mode. It differs from [`ValidiumModeL1BatchCommitDataGenerator`]
+/// in that it includes the pubdata in the produced message.
+#[derive(Debug, Clone)]
+pub struct RollupModeL1BatchCommitDataGenerator;
+
+/// [`ValidiumModeL1BatchCommitDataGenerator`] implements [`L1BatchCommitDataGenerator`] for
+/// contracts operating in validium mode. It differs from [`RollupModeL1BatchCommitDataGenerator`]
+/// in that it does not include the pubdata in the produced message.
+#[derive(Debug, Clone)]
+pub struct ValidiumModeL1BatchCommitDataGenerator;
+
+impl L1BatchCommitDataGenerator for RollupModeL1BatchCommitDataGenerator {
```

Comment:
btw - behavior of these 2 impls is basically identical. 

We should either merge CommitBatchesRollup and CommitBatchesValidium, into one (and then it can internally do the right thing when into_tokens is called), or use generics template here.

Reply:
This was done per our request and this is intentional.